tests/msc2716_test.go

```diff
@@ -0,0 +1,643 @@
+			eventIDsFromResponse = append(eventIDsFromResponse, r.Get("event_id").Str+" ("+r.Get("content").Get("body").Str+")")
+		}
+		return true
+	})
+
+	return eventIDsFromResponse
+}
+
+var txnID int = 0
+
+// The transactions need to be prefixed so they don't collide with the txnID in client.go
+var txnPrefix string = "msc2716-txn"
+
+func sendEvent(t *testing.T, c *client.CSAPI, virtualUserID string, roomID string, e event) string {
+	txnID++
+
+	query := make(url.Values, len(e.PrevEvents))
+	for _, prevEvent := range e.PrevEvents {
+		query.Add("prev_event", prevEvent)
+	}
+
+	if e.OriginServerTS != 0 {
+		query.Add("ts", strconv.FormatUint(e.OriginServerTS, 10))
+	}
+
+	if virtualUserID != "" {
+		query.Add("user_id", virtualUserID)
+	}
+
+	b, err := json.Marshal(e.Content)
+	if err != nil {
+		t.Fatalf("msc2716.sendEvent failed to marshal JSON body: %s", err)
+	}
+
+	res := c.MustDoFunc(t, "PUT", []string{"_matrix", "client", "r0", "rooms", roomID, "send", e.Type, txnPrefix + strconv.Itoa(txnID)}, client.WithRawBody(b), client.WithContentType("application/json"), client.WithQueries(query))
+	body := client.ParseJSON(t, res)
+	eventID := client.GetJSONFieldStr(t, body, "event_id")
+
+	return eventID
+}
+
+// ensureVirtualUserRegistered makes sure the user is registered for the homeserver regardless
+// if they are already registered or not. If unable to register, fails the test
+func ensureVirtualUserRegistered(t *testing.T, c *client.CSAPI, virtualUserLocalpart string) {
+	// b, err := json.Marshal(map[string]interface{}{
+	// 	"username": virtualUserLocalpart,
+	// })
+	// if err != nil {
+	// 	t.Fatalf("msc2716.ensureVirtualUserRegistered failed to marshal JSON body: %s", err)
+	// }
+
+	res := c.DoFunc(
+		t,
+		"POST",
+		[]string{"_matrix", "client", "r0", "register"},
+		client.WithRawBody(json.RawMessage(fmt.Sprintf(`{ "type": "m.login.application_service", "username": "%s" }`, virtualUserLocalpart))),
+		client.WithContentType("application/json"),
+	)
+
+	if res.StatusCode == 200 {
+		return
+	}
+
+	body := client.ParseJSON(t, res)
+	errcode := client.GetJSONFieldStr(t, body, "errcode")
+
+	if res.StatusCode == 400 && errcode == "M_USER_IN_USE" {
+		return
+	} else {
+		errorMessage := client.GetJSONFieldStr(t, body, "error")
+		t.Fatalf("msc2716.ensureVirtualUserRegistered failed to register: (%s) %s", errcode, errorMessage)
+	}
+}
+
+func createMessagesInRoom(t *testing.T, c *client.CSAPI, roomID string, count int) []string {
+	evs := make([]string, count)
+	for i := 0; i < len(evs); i++ {
+		newEvent := b.Event{
+			Type: "m.room.message",
+			Content: map[string]interface{}{
+				"msgtype": "m.text",
+				"body":    fmt.Sprintf("Message %d", i),
+			},
+		}
+		newEventId := c.SendEventSynced(t, roomID, newEvent)
+		evs[i] = newEventId
+	}
+
+	return evs
+}
+
+var chunkCount int64 = 0
+
+func backfillBatchHistoricalMessages(
```

Comment:
Really bad name. It isn't backfilling. It's inserting new events which happen to be old events. Please rename this to be more intuitive.

Reply:
`backfill` seems correct to me. Any event you insert next to an existing event will be back in time around those events?

The [MSC even mentions `backfill` as a noun](https://github.com/matrix-org/matrix-doc/pull/2716/files#diff-2f982a4f06c43a236da1872b37fd632291b5767de9af22084aef85aa4c7a92acR14).

Am I missing a use case where the events aren't old?

Reply:
Federation uses the endpoint `/backfill` to retrieve older messages. I.e paginating backwards in time. This function isn't doing that though, it's _creating_ new events, so it's not clear that `backfillBatchHistoricalMessages` means "make these old events", I intuitively think it's "retrieve older events" (read only)

Reply:
I've renamed it to `batchSendHistoricalMessages` to make it less confusing :+1:

Reply:
Much better name, thank you!!